### PR TITLE
feat(gatsby): improve error messages at runtime

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Adding some better error messages in our runtime code so people can actually see what's going wrong in their error trackers.
It should help us solve https://github.com/gatsbyjs/gatsby/issues/19618

When you have a runtime error in your page component before React kicks in
```js
if (typeof window === 'undefined') {
  throw new Error('GATSBY');
}

export default function MyPage(){}
```
image

When Static Query fails we got an unexpected JSON error, now we get:
image

When Page Query fails, it's still the same:
image

## Related Issues

Related to https://github.com/gatsbyjs/gatsby/issues/19618

### CODE REVIEW

File: e2e-tests/production-runtime/cypress/integration/resource-loading-resilience.js [hunk 1]
```diff
@@ -103,6 +78,12 @@ const runSuiteForPage = (label, pagePath) => {
         filter: `page-data`,
       })
     })
+    describe(`Missing "${label}" static query results`, () => {
```
Comment (wardpeet) [inline e2e-tests/production-runtime/cypress/integration/resource-loading-resilience.js:81]: We weren't tested static-query blocked

File: packages/gatsby/cache-dir/dev-loader.js [hunk 2]
```diff
@@ -25,13 +25,20 @@ function mergePageEntry(cachedPage, newPageData) {
...
+        throw new Error(
+          `We couldn't find the correct component chunk with the name "${chunkName}"`
+        )
+      }
+
+      return (
+        asyncRequires.components[chunkName]()
+          .then(preferDefault)
+          // loader will handle the case when component is error
+          .catch(err => err)
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/dev-loader.js:39]: Return the error instead of throwing

File: packages/gatsby/cache-dir/dev-loader.js [hunk 3]
```diff
@@ -25,13 +25,20 @@ function mergePageEntry(cachedPage, newPageData) {
...
-        ? asyncRequires.components[chunkName]()
-            .then(preferDefault)
-            // loader will handle the case when component is null
-            .catch(() => null)
-        : Promise.resolve()
+    const loadComponent = chunkName => {
+      if (!asyncRequires.components[chunkName]) {
+        throw new Error(
+          `We couldn't find the correct component chunk with the name "${chunkName}"`
+        )
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/dev-loader.js:32]: We silently fail, now we don't

**Comment by wardpeet** [comment]
Probably good to backport this to v2, it will need small change in the cherry-pick PR

**Review by LekoArts — APPROVED** [review]
Overall LGTM. Non blocking (maybe as follow-up): Can we give clearer instructions on how people might solve their issue? We're outputting more detailed errors now, but they are the nitty gritty details. I could imagine a link to a specific docs page for example

**Comment by wardpeet** [comment]
I agree but the issue is that I don't know yet why they happen.

**Pull request merged**